process/blockProcessorCache.go

```diff
@@ -0,0 +1,75 @@
+package process
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+type cacheableBlock interface {
+	Hash() string
+	Nonce() uint64
+}
+
+// No error checks for this cache.
+// These caching errors should never happen, and if they do, they should not be blocking
+
+func (bp *BlockProcessor) cacheObject(obj cacheableBlock, scope string, opts interface{}) {
+	objKey := makeObjKey(scope, obj.Hash(), opts)
+	hashLookupKey := makeHashCacheKey(scope, obj.Hash(), opts)
+	nonceLookupKey := makeNonceCacheKey(scope, obj.Nonce(), opts)
+
+	// Store object
+	_ = bp.cache.Put(objKey, obj, 0)
+
+	// Store nonce + hash lookup keys
+	_ = bp.cache.Put(hashLookupKey, objKey, 0)
+	_ = bp.cache.Put(nonceLookupKey, objKey, 0)
+}
+
+func makeObjKey(scope string, hash string, opts interface{}) []byte {
+	optBytes, _ := json.Marshal(opts)
+	return []byte(fmt.Sprintf("%s:%s|opts:%s", scope, hash, string(optBytes)))
+}
+
+func makeHashCacheKey(scope string, hash string, opts interface{}) []byte {
+	optBytes, _ := json.Marshal(opts)
+	return []byte(fmt.Sprintf("%s:hash:%s|opts:%s", scope, hash, string(optBytes)))
+}
+
+func makeNonceCacheKey(scope string, nonce uint64, opts interface{}) []byte {
+	optBytes, _ := json.Marshal(opts)
```

Comment:
in case of error here, I would log the error, and then simply not cache.
the fallback in case something is not cached, will be to take it from the observer.